File: src/nac/basic/source_location.cpp
#include "sodium/nac/basic/source_location.h"

#include "sodium/nac/basic/source_range.h"

namespace sodium {

SourceLocation::SourceLocation() : line_(0), column_(0), position_(nullptr) {}

SourceLocation::SourceLocation(int line, int column, const char *position)
        : line_(line),
          column_(column),
          position_(position) {}

int SourceLocation::line() const {
    return line_;
}

int SourceLocation::column() const {
    return column_;
}

const char *SourceLocation::position() const {
    return position_;
}

bool SourceLocation::is_valid() const {
    return line_ > 0 && column_ > 0 && position_ != nullptr;
}

SourceRange SourceLocation::to(SourceLocation location) const {
    return SourceRange(*this, location);
}

std::string SourceLocation::to_string() const {
    return std::string(std::to_string(line_) + ":" + std::to_string(column_));
}

} // namespace sodium

File: nac/src/basic/source_location.rs
use std::fmt::Display;

use super::source_range::SourceRange;

#[derive(Clone, Copy)]
pub struct SourceLocation {
    line: u32,
    column: u32,
}

impl SourceLocation {
    pub fn new(line: u32, column: u32) -> Self {
        Self { line, column }
    }

    pub fn line(&self) -> u32 {
        self.line
    }

    pub fn column(&self) -> u32 {
        self.column
    }

    pub fn to(&self, location: &Self) -> SourceRange {
        SourceRange::new(self.clone(), location.clone())
    }
}

impl Display for SourceLocation {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}:{}", self.line, self.column)
    }
}
